### PR TITLE
Add support for multi-imp requests in ttx bidder

### PR DESCRIPTION
**Pull request by snahornyi**
Ported from https://github.com/prebid/prebid-server/pull/1609
except `request.ext.ttx.ATTR` field since it is impact core which shouldn't be happen for bidder's modification.

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java [hunk 1]
```diff
@@ -56,44 +60,68 @@ public TtxBidder(String endpointUrl, JacksonMapper mapper) {
...
-        }
-
-        if (updatedFirstImp != null && updatedFirstImp.getBanner() == null && updatedFirstImp.getVideo() == null) {
-            return Result.withError(BidderError.badInput("At least one of [banner, video] "
-                    + "formats must be defined in Imp. None found"));
-        }
+        final List<HttpRequest<BidRequest>> requests = new ArrayList<>();
 
-        final HttpRequest<BidRequest> httpRequest = createRequest(request, extImpTtx, updatedFirstImp);
+        final ExtRequest reqExt = updateExtRequest(request.getExt());
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java:66]: Catch exception

File: src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java [hunk 2]
```diff
@@ -56,44 +60,68 @@ public TtxBidder(String endpointUrl, JacksonMapper mapper) {
...
 
-        return Result.of(Collections.singletonList(httpRequest), errors);
+        for (Imp imp : request.getImp()) {
+            try {
+                requests.add(makeRequest(request, imp, reqExt));
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java:70]: makeRequest seems redundant. 
Could you split it like. 
validateImp()
parseImpExt()
updateImp()
 createRequest()

File: src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java [hunk 3]
```diff
@@ -132,14 +159,15 @@ private static Integer resolvePlacement(Integer videoPlacement, String productId
...
+    private ObjectNode createImpExt(String productId, String zoneId, String siteId) {
+        final TtxImpExt ttxImpExt = TtxImpExt.of(
+                TtxImpExtTtx.of(productId, StringUtils.isNotEmpty(zoneId) ? zoneId : siteId));
         return mapper.mapper().valueToTree(ttxImpExt);
     }
 
-    private HttpRequest<BidRequest> createRequest(BidRequest request, ExtImpTtx extImpTtx, Imp updatedFirstImp) {
-        final Site updatedSite = extImpTtx != null ? updateSite(request.getSite(), extImpTtx.getSiteId()) : null;
-        final BidRequest modifiedRequest = updateRequest(request, updatedSite, updatedFirstImp);
+    private HttpRequest<BidRequest> createRequest(BidRequest request,
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java:168]: redundant /n

File: src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java [hunk 4]
```diff
@@ -150,27 +178,14 @@ private ObjectNode createImpExt(String productId, String zoneId) {
                 .build();
     }
 
-    private static Site updateSite(Site site, String siteId) {
-        return site == null ? null : site.toBuilder().id(siteId).build();
-    }
+    private static BidRequest updateRequest(BidRequest request, Imp requestImp, ExtRequest extRequest) {
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java:181]: seems redundant method

**Review by DGarbar — CHANGES_REQUESTED** [review]

**Review by DGarbar — DISMISSED** [review]

File: src/main/java/org/prebid/server/proto/openrtb/ext/FlexibleExtension.java [hunk 5]
```diff
@@ -19,7 +20,15 @@
...
-    private final Map<String, JsonNode> properties = new HashMap<>();
+    private final Map<String, JsonNode> properties;
+
+    protected FlexibleExtension(Map<String, JsonNode> properties) {
+        this.properties = MapUtils.isNotEmpty(properties) ? new HashMap<>(properties) : new HashMap<>();
+    }
+
+    protected FlexibleExtension() {
+        this.properties = new HashMap<>();
+    }
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/proto/openrtb/ext/FlexibleExtension.java:31]: Let's discuss this approach

**Review by rpanchyk — DISMISSED** [review]

**Review by rpanchyk — APPROVED** [review]

**Pull request merged**